dbos/_core/system_database.py

```diff
@@ -0,0 +1,1734 @@
+        with self.sync_engine.begin() as c:
+            init_recv = (c.execute(get_sql)).fetchall()
+
+        value: Any = None
+        if len(init_recv) > 0:
+            value = serialization.deserialize(init_recv[0][0])
+        else:
+            # Wait for the notification
+            actual_timeout = timeout_seconds
+            if caller_ctx is not None:
+                # Support OAOO sleep for workflows
+                actual_timeout = self.sleep_sync(
+                    caller_ctx["workflow_uuid"],
+                    caller_ctx["timeout_function_id"],
+                    timeout_seconds,
+                    skip_sleep=True,
+                )
+            condition.wait(timeout=actual_timeout)
+
+            # Read the value from the database
+            with self.sync_engine.begin() as c:
+                final_recv = (c.execute(get_sql)).fetchall()
+                if len(final_recv) > 0:
+                    value = serialization.deserialize(final_recv[0][0])
+        condition.release()
+        self.workflow_events_map.pop(payload)
+
+        # Record the output if it's in a workflow
+        if caller_ctx is not None:
+            self.record_operation_result_sync(
+                {
+                    "workflow_uuid": caller_ctx["workflow_uuid"],
+                    "function_id": caller_ctx["function_id"],
+                    "output": serialization.serialize(
+                        value
+                    ),  # None will be serialized to 'null'
+                    "error": None,
+                }
+            )
+        return value
+
+    async def get_event_async(
+        self,
+        target_uuid: str,
+        key: str,
+        timeout_seconds: float = 60,
+        caller_ctx: Optional[GetEventWorkflowContext] = None,
+    ) -> Any:
+        get_sql = sa.select(
+            SystemSchema.workflow_events.c.value,
+        ).where(
+            SystemSchema.workflow_events.c.workflow_uuid == target_uuid,
+            SystemSchema.workflow_events.c.key == key,
+        )
+        # Check for previous executions only if it's in a workflow
+        if caller_ctx is not None:
+            recorded_output = await self.check_operation_execution_async(
+                caller_ctx["workflow_uuid"], caller_ctx["function_id"]
+            )
+            if recorded_output is not None:
+                dbos_logger.debug(
+                    f"Replaying get_event, id: {caller_ctx['function_id']}, key: {key}"
+                )
+                if recorded_output["output"] is not None:
+                    return serialization.deserialize(recorded_output["output"])
+                else:
+                    raise Exception("No output recorded in the last get_event")
+            else:
+                dbos_logger.debug(
+                    f"Running get_event, id: {caller_ctx['function_id']}, key: {key}"
+                )
+
+        payload = f"{target_uuid}::{key}"
+        condition = threading.Condition()
```

Comment:
asyncio.Condition?

Reply:
I was worried about having different condition classes for sync/async versions. I will investigate further